src/charm.py

```diff
@@ -448,6 +506,30 @@ def is_db_service_ready(self) -> bool:
         with MongoDBConnection(self.mongodb_config, "localhost", direct=True) as direct_mongo:
             return direct_mongo.is_ready
 
+    def _on_config_changed(self, event: ConfigChangedEvent) -> None:
+        """Listen to changes in application configuration.
+
+        To prevent a user from migrating a cluster, and causing the component to become
+        unresponsive therefore causing a cluster failure, error the component. This prevents it
+        from executing other hooks with a new role.
+        """
+        if self.is_role_changed():
```

Comment:
nit: (to reduce one layer of nesting)
```suggestion
        if not self.is_role_changed():
            return
```